cpp/tests/io/json_whitespace_normalization_test.cu

```diff
@@ -0,0 +1,207 @@
+
+// The DFA's starting state
+constexpr char start_state = static_cast<char>(TT_OOS);
+
+struct TransduceToNormalizedWS {
+  /**
+   * @brief Returns the <relative_offset>-th output symbol on the transition (state_id, match_id).
+   */
+  template <typename StateT, typename SymbolGroupT, typename RelativeOffsetT, typename SymbolT>
+  constexpr CUDF_HOST_DEVICE SymbolT operator()(StateT const state_id,
+                                                SymbolGroupT const match_id,
+                                                RelativeOffsetT const relative_offset,
+                                                SymbolT const read_symbol) const
+  {
+    // -------- TRANSLATION TABLE ------------
+    //      Let the alphabet set be Sigma
+    // ---------------------------------------
+    // ---------- NON-SPECIAL CASES: ----------
+    //      Output symbol same as input symbol <s>
+    // state | read_symbol <s>  -> output_symbol <s>
+    // DQS   | Sigma            -> Sigma
+    // WS    | Sigma\{<SPC>,\t} -> Sigma\{<SPC>,\t}
+    // OOS   | Sigma\{<SPC>,\t} -> Sigma\{<SPC>,\t}
+    // ---------- SPECIAL CASES: --------------
+    //    Input symbol translates to output symbol
+    // OOS   | {<SPC>}          -> <nop>
+    // OOS   | {\t}             -> <nop>
+    // WS    | {<SPC>}          -> <nop>
+    // WS    | {\t}             -> <nop>
+
+    // Case when read symbol is a whitespace or tabspace but is unquoted
+    if (((match_id == static_cast<SymbolGroupT>(dfa_symbol_group_id::WHITESPACE_CHAR)) ||
+         (match_id == static_cast<SymbolGroupT>(dfa_symbol_group_id::TABSPACE_CHAR))) &&
+        ((state_id == static_cast<StateT>(dfa_states::TT_OOS)) ||
+         (state_id == static_cast<StateT>(dfa_states::TT_WS)))) {
+      return 0;
+    }
+    // In all other cases we simply output the input symbol
+    return read_symbol;
+  }
+
+  /**
+   * @brief Returns the number of output characters for a given transition.
+   * During quote normalization, we always emit one output character (i.e., either the input
+   * character or the single quote-input replaced by a double quote), except when we need to escape
+   * a double quote that was previously inside a single-quoted string.
+   */
+  template <typename StateT, typename SymbolGroupT, typename SymbolT>
+  constexpr CUDF_HOST_DEVICE int32_t operator()(StateT const state_id,
+                                                SymbolGroupT const match_id,
+                                                SymbolT const read_symbol) const
+  {
+    // Case when read symbol is a whitespace or tabspace but is unquoted
+    if (((match_id == static_cast<SymbolGroupT>(dfa_symbol_group_id::WHITESPACE_CHAR)) ||
+         (match_id == static_cast<SymbolGroupT>(dfa_symbol_group_id::TABSPACE_CHAR))) &&
+        ((state_id == static_cast<StateT>(dfa_states::TT_OOS)) ||
+         (state_id == static_cast<StateT>(dfa_states::TT_WS)))) {
```

Comment:
This relates to my above comment on using a symbol group for `space` and `tab`.
I'm not 100% I got the parentheses right, could you please double check? 🙏 
```suggestion
    if ((match_id == static_cast<SymbolGroupT>(dfa_symbol_group_id::WHITESPACE_CHAR)) &&
        ((state_id == static_cast<StateT>(dfa_states::TT_OOS)) ||
         (state_id == static_cast<StateT>(dfa_states::TT_WS)))) {
```